native/components/message-markdown-rules.react.js

```diff
@@ -0,0 +1,136 @@
+import * as SimpleMarkdown from 'simple-markdown';
+
+export default function rules(
+  styles: StyleSheetOf<Style>,
+  onLink: (url: string) => Promise<any>,
+) {
+  const LINK_INSIDE = '(?:\\[[^\\]]*\\]|[^\\]]|\\](?=[^\\[]*\\]))*';
+  const LINK_HREF_AND_TITLE =
+    '\\s*<?([^\\s]*?)>?(?:\\s+[\'"]([\\s\\S]*?)[\'"])?\\s*';
+  const pressHandler = target => {
+    if (onLink) {
+      onLink(target).catch(error => {
+        console.log(
+          'There has been a problem with this action. ' + error.message,
+        );
+        throw error;
+      });
+    }
+  };
+
+  return {
+    autolink: {
+      ...SimpleMarkdown.defaultRules.autolink,
+      react: (
+        node: SimpleMarkdown.SingleASTNode,
+        output: SimpleMarkdown.Output<SimpleMarkdown.ReactElement>,
+        state: SimpleMarkdown.State,
+      ) => {
+        state.withinText = true;
+        const _pressHandler = () => {
+          pressHandler(node.target);
+        };
+        const autolink = (
+          <Text key={state.key} style={styles.autolink} onPress={_pressHandler}>
+            {output(node.content, state)}
+          </Text>
+        );
+        return autolink;
+      },
+    },
+    link: {
+      ...SimpleMarkdown.defaultRules.link,
+      match: SimpleMarkdown.inlineRegex(
+        new RegExp(
+          '^\\[(' + LINK_INSIDE + ')\\]\\(' + LINK_HREF_AND_TITLE + '\\)',
+        ),
+      ),
+      react: (
+        node: SimpleMarkdown.SingleASTNode,
+        output: SimpleMarkdown.Output<SimpleMarkdown.ReactElement>,
+        state: SimpleMarkdown.State,
+      ) => {
+        state.withinLink = true;
+        const _pressHandler = () => {
+          pressHandler(node.target);
+        };
+        const link = (
+          <Text key={state.key} style={styles.autolink} onPress={_pressHandler}>
+            {output(node.content, state)}
+          </Text>
+        );
+        state.withinLink = false;
+        return link;
+      },
+    },
+    // in some cases, simple-markdown recognizes "normal" text as a paragraph
+    // and we need this rule to preserve our style and bubble size
+    paragraph: {
+      ...SimpleMarkdown.defaultRules.paragraph,
+      react: (
+        node: SimpleMarkdown.SingleASTNode,
+        output: SimpleMarkdown.Output<SimpleMarkdown.ReactElement>,
+        state: SimpleMarkdown.State,
+      ) => {
+        const paragraph = (
+          <Text key={state.key} style={styles.paragraph}>
+            {output(node.content, state)}
+          </Text>
+        );
+        return paragraph;
+      },
+    },
+    // rule to preserve our text style in simple messages
+    text: {
+      ...SimpleMarkdown.defaultRules.text,
+      react: (
+        node: SimpleMarkdown.SingleASTNode,
+        output: SimpleMarkdown.Output<string>,
+        state: SimpleMarkdown.State,
+      ) => {
+        let textStyle = {
+          ...styles.text,
+          ...(state.style || {}),
+        };
+
+        if (state.withinLink) {
```

Comment:
That's interesting, didn't realize you could have a `text` node within a `link` node. I'd like to understand how `SimpleMarkdown`'s parser works better. Can you give me an example of Markdown syntax that will trigger `withinLink` here, and the corresponding AST that `SimpleMarkdown`'s parser generates from it?

Reply:
Basically, every `link` node contains `text` node inside. Content in `text` is equal to target in `link` 
![Zrzut ekranu 2020-07-7 o 11 20 33](https://user-images.githubusercontent.com/18313471/86758107-e8d77e80-c043-11ea-95b3-5074e2f736ae.png)